### PR TITLE
Add Angular explorer integration tests

### PR DESCRIPTION
**Pull request by josemontespg**
This PR adds a new suite of integration tests for the Angular explorer, exercising the full pipeline from a JSON A2UI payload to the rendered DOM.

Also, added these tests to the CI, to run automatically on every commit.

## What
These tests act as integration tests that verify all the code between a JSON A2UI payload and the rendered DOM. 
This includes core_web, angular renderer, and the explorer demo app.
The tests use standard DOM queries to check expectations, which is more likely to be stable over time than screenshot/golden tests. Some tests are interactive, and click buttons, type on input boxes, open modals, etc.

## Why
The motivation for these tests is http://b/511068435#comment3 because we were missing a test for the modal component. I found that we were missing tests for the explorer app, which is already a great sample app to exercise many different features and edge cases.

### CODE REVIEW

File: renderers/angular/a2ui_explorer/src/app/tests/test_utils.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  fixture.detectChanges();
+
+  return fixture;
+}
+
+/**
+ * Helper function to wait for a given number of milliseconds.
+ */
+export function wait(ms: number) {
+  return new Promise(resolve => setTimeout(resolve, ms));
```
Comment (andrewkolos) [inline renderers/angular/a2ui_explorer/src/app/tests/test_utils.ts:65]: These tests are super flaky on my machine.

In my experience working on the flutter/flutter repo, timeout waits can be a huge source of test flakiness and/or wasted wall time. Unfortunately, I imagine a properly deterministic solution would be involve a complex change to the library given the amount of recursive reactivity involved.

File: renderers/angular/a2ui_explorer/src/app/tests/02_email-compose.spec.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  });
+
+  it('should handle Send button click', async () => {
+    const buttons = [
+      ...fixture.nativeElement.querySelectorAll('.a2ui-button'),
+    ] as HTMLButtonElement[];
+    const sendBtn = buttons.find(b => b.textContent.includes('Send'))!;
+    expect(sendBtn).withContext('Should find Send button').toBeTruthy();
+
+    (sendBtn as any).click();
```
Comment (andrewkolos) [inline renderers/angular/a2ui_explorer/src/app/tests/02_email-compose.spec.ts]: 1. avoid `any`, cast to `HTMLButtonElement` if needed.
2. `sendBtn` already has a type of `HTMLButtonElement` thanks to the earlier cast on `buttons`.
3. this can be prevented with linting (`@typescript-eslint/no-unnecessary-type-assertion`. Filed https://github.com/google/A2UI/issues/1387

File: renderers/angular/a2ui_explorer/src/app/tests/test_utils.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  const example = EXAMPLES.find(ex => ex.name === exampleName);
+  expect(example).toBeTruthy();
+
+  component.selectExample(example!);
+  fixture.detectChanges();
+
+  // Wait for async operations (like agent stub playback) to complete
+  await wait(50);
+  fixture.detectChanges();
```
Comment (andrewkolos) [inline renderers/angular/a2ui_explorer/src/app/tests/test_utils.ts]: This wait appears to be too short for tests to pass locally.

File: renderers/angular/a2ui_explorer/src/app/tests/02_email-compose.spec.ts (hunk 2 above)
Comment (josemontespg) [reply renderers/angular/a2ui_explorer/src/app/tests/02_email-compose.spec.ts]: Good catch, thank you! I agree the linter should catch these kinds of things. I'll work on enabling the linter next

File: renderers/angular/a2ui_explorer/src/app/tests/test_utils.ts (hunk 1 above)
Comment (josemontespg) [reply renderers/angular/a2ui_explorer/src/app/tests/test_utils.ts:65]: I tried multiple approaches, and could not find a way to cleanly detect when the component under test is stable. I think the source of complexity is the recursive signal listeners for each of the data model paths. I switched to a different approach of waiting for a number of macrotasks, instead of waiting for a ms amount, which I think should be more stable.

File: renderers/angular/a2ui_explorer/src/app/tests/test_utils.ts (hunk 3 above)
Comment (josemontespg) [reply renderers/angular/a2ui_explorer/src/app/tests/test_utils.ts]: I changed the approach to wait for a number of macrotasks instead of a number of milliseconds. It's very stable on my end now.

**Comment by ditman** [comment]
When running these tests on my linux box, it seems that the process hangs at the end after all tests have passed. Is this a problem with my machine or the test setup? I think I've seen other karma tests behave similarly in my box :/

**Comment by andrewkolos** [comment]
> When running these tests on my linux box, it seems that the process hangs at the end after all tests have passed. Is this a problem with my machine or the test setup? I think I've seen other karma tests behave similarly in my box :/

FWIW they seem to run fine on my linux VM:

<details>

<summary> command output </summary>

```
cd /renderers/angular
andrewkolos@andrewkolos-linux:~/Documents/A2UI/renderers/angular$ npx ng test a2ui_explorer --watch=false --browsers=ChromeHeadless
npm warn Unknown project config "always-auth" (//us-npm.pkg.dev/oss-exit-gate-prod/a2ui--npm/:always-auth). This will stop working in the next major version of npm. See `npm help npmrc` for supported config options.
Initial chunk files                               | Names                                          |  Raw size
chunk-LUAA3RTM.js                                 | -                                              |   2.31 MB | 
polyfills.js                                      | polyfills                                      | 921.33 kB | 
chunk-6U3DUBJX.js                                 | -                                              | 530.11 kB | 
chunk-CON7GZBZ.js                                 | -                                              | 388.01 kB | 
test_main.js                                      | test_main                                      |  56.68 kB | 
chunk-YFCGYO6H.js                                 | -                                              |   6.24 kB | 
spec-app-tests-30_live-invitation-builder.spec.js | spec-app-tests-30_live-invitation-builder.spec |   5.22 kB | 
spec-app-tests-32_advanced-form-validator.spec.js | spec-app-tests-32_advanced-form-validator.spec |   3.33 kB | 
spec-app-tests-25_contact-card.spec.js            | spec-app-tests-25_contact-card.spec            |   3.25 kB | 
spec-app-tests-13_coffee-order.spec.js            | spec-app-tests-13_coffee-order.spec            |   2.97 kB | 
spec-app-agent-stub.service.spec.js               | spec-app-agent-stub.service.spec               |   2.96 kB | 
spec-app-tests-19_software-purchase.spec.js       | spec-app-tests-19_software-purchase.spec       |   2.93 kB | 
spec-app-tests-02_email-compose.spec.js           | spec-app-tests-02_email-compose.spec           |   2.92 kB | 
spec-app-tests-17_event-detail.spec.js            | spec-app-tests-17_event-detail.spec            |   2.91 kB | 
spec-app-tests-24_recipe-card.spec.js             | spec-app-tests-24_recipe-card.spec             |   2.89 kB | 
spec-app-tests-15_account-balance.spec.js         | spec-app-tests-15_account-balance.spec         |   2.70 kB | 
spec-app-tests-10_notification-permission.spec.js | spec-app-tests-10_notification-permission.spec |   2.69 kB | 
spec-app-tests-29_movie-card.spec.js              | spec-app-tests-29_movie-card.spec              |   2.66 kB | 
spec-app-tests-11_purchase-complete.spec.js       | spec-app-tests-11_purchase-complete.spec       |   2.42 kB | 
spec-app-tests-18_track-list.spec.js              | spec-app-tests-18_track-list.spec              |   2.34 kB | 
spec-app-tests-05_product-card.spec.js            | spec-app-tests-05_product-card.spec            |   2.22 kB | 
spec-app-tests-07_task-card.spec.js               | spec-app-tests-07_task-card.spec               |   2.05 kB | 
spec-app-tests-04_weather-current.spec.js         | spec-app-tests-04_weather-current.spec         |   2.02 kB | 
spec-app-tests-03_calendar-day.spec.js            | spec-app-tests-03_calendar-day.spec            |   1.93 kB | 
spec-app-app.spec.js                              | spec-app-app.spec                              |   1.90 kB | 
spec-app-tests-26_podcast-episode.spec.js         | spec-app-tests-26_podcast-episode.spec         |   1.88 kB | 
spec-app-tests-12_chat-message.spec.js            | spec-app-tests-12_chat-message.spec            |   1.84 kB | 
spec-app-tests-08_user-profile.spec.js            | spec-app-tests-08_user-profile.spec            |   1.81 kB | 
spec-app-tests-20_restaurant-card.spec.js         | spec-app-tests-20_restaurant-card.spec         |   1.76 kB | 
spec-app-tests-33_financial-data-grid.spec.js     | spec-app-tests-33_financial-data-grid.spec     |   1.76 kB | 
spec-app-tests-06_music-player.spec.js            | spec-app-tests-06_music-player.spec            |   1.66 kB | 
spec-app-tests-01_flight-status.spec.js           | spec-app-tests-01_flight-status.spec           |   1.66 kB | 
spec-app-tests-14_sports-player.spec.js           | spec-app-tests-14_sports-player.spec           |   1.65 kB | 
chunk-Z5GBGFMC.js                                 | -                                              |   1.64 kB | 
spec-app-tests-21_shipping-status.spec.js         | spec-app-tests-21_shipping-status.spec         |   1.60 kB | 
spec-app-tests-31_incremental-dashboard.spec.js   | spec-app-tests-31_incremental-dashboard.spec   |   1.57 kB | 
spec-app-tests-36_modal.spec.js                   | spec-app-tests-36_modal.spec                   |   1.53 kB | 
spec-app-tests-16_workout-summary.spec.js         | spec-app-tests-16_workout-summary.spec         |   1.50 kB | 
spec-app-tests-34_child-list-template.spec.js     | spec-app-tests-34_child-list-template.spec     |   1.49 kB | 
spec-app-tests-22_credit-card.spec.js             | spec-app-tests-22_credit-card.spec             |   1.48 kB | 
spec-app-tests-28_countdown-timer.spec.js         | spec-app-tests-28_countdown-timer.spec         |   1.41 kB | 
spec-app-tests-23_step-counter.spec.js            | spec-app-tests-23_step-counter.spec            |   1.27 kB | 
spec-app-tests-09_login-form.spec.js              | spec-app-tests-09_login-form.spec              |   1.18 kB | 
spec-app-tests-35_markdown-text.spec.js           | spec-app-tests-35_markdown-text.spec           |   1.16 kB | 
spec-app-tests-27_stats-card.spec.js              | spec-app-tests-27_stats-card.spec              |   1.03 kB | 
jasmine-cleanup-0.js                              | jasmine-cleanup-0                              | 519 bytes | 

                                                  | Initial total                                  |   4.30 MB

Lazy chunk files                                  | Names                                          |  Raw size
chunk-BBBXSZ7F.js                                 | markdown                                       | 313.51 kB | 

Application bundle generation complete. [4.653 seconds] - 2026-05-11T21:51:58.195Z

▲ [WARNING] TypeScript compiler options 'module' values 'CommonJS', 'UMD', 'System' and 'AMD' are not supported. [plugin angular-compiler]

  The 'module' option will be set to 'ES2022' instead.


11 05 2026 21:51:58.579:INFO [karma-server]: Karma v6.4.4 server started at http://127.0.0.1:9876/
11 05 2026 21:51:58.580:INFO [launcher]: Launching browsers ChromeHeadless with concurrency unlimited
11 05 2026 21:51:58.584:INFO [launcher]: Starting browser ChromeHeadless
11 05 2026 21:52:00.501:INFO [Chrome Headless 148.0.0.0 (Linux 0.0.0)]: Connected on socket 4I88vgN4SJvM9MdgAAAB with id 75198364
LOG: '[AgentStub] handleAction action:', Object{name: 'call', surfaceId: 'gallery-contact-card', sourceComponentId: 'call-btn', timestamp: '2026-05-11T21:52:02.559Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 5 of 121 SUCCESS (0 secs / 0.87 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'call', surfaceId: 'gallery-contact-card', sour
LOG: '[AgentStub] handleAction action:', Object{name: 'message', surfaceId: 'gallery-contact-card', sourceComponentId: 'message-btn', timestamp: '2026-05-11T21:52:02.853Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 7 of 121 SUCCESS (0 secs / 1.157 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'message', surfaceId: 'gallery-contact-card', s
LOG: '[AgentStub] handleAction action:', Object{name: 'add_to_cart', surfaceId: 'gallery-coffee-order', sourceComponentId: 'add-btn', timestamp: '2026-05-11T21:52:03.045Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 8 of 121 SUCCESS (0 secs / 1.32 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'add_to_cart', surfaceId: 'gallery-coffee-order
LOG: '[AgentStub] handleAction action:', Object{name: 'purchase', surfaceId: 'gallery-coffee-order', sourceComponentId: 'purchase-btn', timestamp: '2026-05-11T21:52:03.393Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 10 of 121 SUCCESS (0 secs / 1.647 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'purchase', surfaceId: 'gallery-coffee-order', 
LOG: '[AgentStub] handleAction action:', Object{name: 'view_details', surfaceId: 'gallery-purchase-complete', sourceComponentId: 'view-btn', timestamp: '2026-05-11T21:52:04.447Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 16 of 121 SUCCESS (0 secs / 2.637 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'view_details', surfaceId: 'gallery-purchase-co
LOG: '[AgentStub] handleAction action:', Object{name: 'decline', surfaceId: 'gallery-notification-permission', sourceComponentId: 'no-btn', timestamp: '2026-05-11T21:52:05.526Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 23 of 121 SUCCESS (0 secs / 3.749 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'decline', surfaceId: 'gallery-notification-per
LOG: '[AgentStub] handleAction action:', Object{name: 'accept', surfaceId: 'gallery-notification-permission', sourceComponentId: 'yes-btn', timestamp: '2026-05-11T21:52:05.743Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 24 of 121 SUCCESS (0 secs / 3.921 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'accept', surfaceId: 'gallery-notification-perm
LOG: '[AgentStub] handleAction action:', Object{name: 'send', surfaceId: 'gallery-email-compose', sourceComponentId: 'send-btn', timestamp: '2026-05-11T21:52:07.906Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 38 of 121 SUCCESS (0 secs / 6.094 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'send', surfaceId: 'gallery-email-compose', sou
LOG: '[AgentStub] handleAction action:', Object{name: 'discard', surfaceId: 'gallery-email-compose', sourceComponentId: 'discard-btn', timestamp: '2026-05-11T21:52:08.103Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 39 of 121 SUCCESS (0 secs / 6.254 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'discard', surfaceId: 'gallery-email-compose', 
LOG: '[AgentStub] handleAction action:', Object{name: 'decline', surfaceId: 'gallery-event-detail', sourceComponentId: 'decline-btn', timestamp: '2026-05-11T21:52:08.818Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 45 of 121 SUCCESS (0 secs / 6.988 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'decline', surfaceId: 'gallery-event-detail', s
LOG: '[AgentStub] handleAction action:', Object{name: 'accept', surfaceId: 'gallery-event-detail', sourceComponentId: 'accept-btn', timestamp: '2026-05-11T21:52:09.171Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 48 of 121 SUCCESS (0 secs / 7.341 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'accept', surfaceId: 'gallery-event-detail', so
LOG: '[AgentStub] handleAction action:', Object{name: 'confirm', surfaceId: 'gallery-software-purchase', sourceComponentId: 'confirm-btn', timestamp: '2026-05-11T21:52:09.589Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 51 of 121 SUCCESS (0 secs / 7.741 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'confirm', surfaceId: 'gallery-software-purchas
LOG: '[AgentStub] handleAction action:', Object{name: 'cancel', surfaceId: 'gallery-software-purchase', sourceComponentId: 'cancel-btn', timestamp: '2026-05-11T21:52:09.753Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 52 of 121 SUCCESS (0 secs / 7.901 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'cancel', surfaceId: 'gallery-software-purchase
LOG: '[AgentStub] handleAction action:', Object{name: 'addToCart', surfaceId: 'gallery-product-card', sourceComponentId: 'add-cart-btn', timestamp: '2026-05-11T21:52:14.738Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 87 of 121 SUCCESS (0 secs / 12.821 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'addToCart', surfaceId: 'gallery-product-card',
WARN: '[DefaultMarkdownRenderer] Failed to load optional `@a2ui/markdown-it` renderer. Using fallback.'
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 102 of 121 SUCCESS (0 secs / 14.963 secs)
WARN: '[DefaultMarkdownRenderer] Failed to load optional `@a2ui/markdown-it` renderer. Using fallback
LOG: '[AgentStub] handleAction action:', Object{name: 'pay_bill', surfaceId: 'gallery-account-balance', sourceComponentId: 'pay-btn', timestamp: '2026-05-11T21:52:17.416Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 107 of 121 SUCCESS (0 secs / 15.444 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'pay_bill', surfaceId: 'gallery-account-balance
LOG: '[AgentStub] handleAction action:', Object{name: 'transfer', surfaceId: 'gallery-account-balance', sourceComponentId: 'transfer-btn', timestamp: '2026-05-11T21:52:17.747Z', context: Object{}}
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 110 of 121 SUCCESS (0 secs / 15.768 secs)
LOG: '[AgentStub] handleAction action:', Object{name: 'transfer', surfaceId: 'gallery-account-balance
Chrome Headless 148.0.0.0 (Linux 0.0.0): Executed 121 of 121 SUCCESS (17.071 secs / 16.692 secs)
TOTAL: 121 SUCCESS
andrewkolos@andrewkolos-linux:~/Documents/A2UI/renderers/angular$ 
```

</details>

**Comment by josemontespg** [comment]
> When running these tests on my linux box, it seems that the process hangs at the end after all tests have passed. Is this a problem with my machine or the test setup? I think I've seen other karma tests behave similarly in my box :/

What command are you using to run the test? There is a `watch` flag that will do what you describe.

If you think the PR is ready, please approve. We can always tweak the code afterwards.

**Comment by ditman** [comment]
I was running `npm run test` from the `renderers/angular` directory :)

(which is NOT the command I was looking for. Let's make a top-level `test:integration` script or similar so people don't have to hunt for specific commands to run the tests!)

**Review by ditman — APPROVED** [review]
Let's go!
